### PR TITLE
Add qr code for visit request

### PR DESCRIPTION
**Pull request by denys-medynskyi**
Resolves [561](https://github.com/pivorakmeetup/pivorak-web-app/issues/561)

### Description

- Sends QR code in an email after user confirms his visit
- Allows admin to scan and check in user via QR code

### Pre-merge checklist
 
- [x] The PR relates to a single subject with a clear title and description
- [x] Verify that feature branch is up-to-date with `development` (if not - rebase it).

### CODE REVIEW

File: app/views/admin/visit_request/check_in/already_checked_in.slim [hunk 1]
```diff
@@ -0,0 +1,9 @@
+.check_in
+  p
+    i.empty.user.icon.green
+    | Name: <b>#{user.first_name}</b> #{user.last_name}
```
Comment (sveredyuk) [inline app/views/admin/visit_request/check_in/already_checked_in.slim]: We try keep all locales in `en.yml` files

File: app/controllers/admin/visit_request/check_in_controller.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
+module Admin
+  class VisitRequest
+    class CheckInController < ::Admin::BaseController
+      helper_method :event, :upcoming_event, :user, :visit_request, :visit_count
+
+      def show
+        begin
+          ::VisitRequest::CheckIn.new(visit_request).call
+          render :show
+        rescue ActiveRecord::RecordNotFound
```
Comment (sveredyuk) [inline app/controllers/admin/visit_request/check_in_controller.rb]: looks to defensife. Maybe to not use exceptions for control flow? Just `:symbols` as status?

Comment (sveredyuk) [reply app/controllers/admin/visit_request/check_in_controller.rb]:
```ruby
def show
  # Just return 2 values from this service. 1st as visit_request object and 2nd as symbol status
  # in such way your controller would not know so much about logic
  @visit_request, status = ::VisitRequest::CheckIn.new(visit_request).call
  render status
end
```

File: app/services/visit_request/check_in.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    InvalidStatus = Class.new(StandardError)
+
+    def initialize(visit_request)
+      @visit_request = visit_request
+    end
+
+    def call
+      raise AlreadyCheckedIn if visit_request.visited?
+      raise InvalidStatus unless visit_request.approved?
+
```
Comment (sveredyuk) [inline app/services/visit_request/check_in.rb:19]:
```ruby
def call
  status = if visit_request.visited?
   :already_visited
  elsif  !visit_request.approved?
    :invalid_status
  else
   :show
end

  visit_request = visit_request.update(visited: true, checked_in_at: Time.zone.now) if status == :show
  [visit_request, status]
end
```

File: spec/services/visit_request/generate_qr_code_spec.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+describe VisitRequest::GenerateQrCode do
+  let(:visit_request) { create(:visit_request).reload }
+
+  subject(:service) { described_class.new(visit_request) }
+
+  describe '#call' do
+    subject(:call) { service.call }
+
+    it 'returns PDF file with QR code' do
+      expect(call).to be_present
```
Comment (sveredyuk) [inline spec/services/visit_request/generate_qr_code_spec.rb]: this spec should compare generated file with pre-generated QR code file as factory

**Review by sveredyuk — COMMENTED** [review]
Well done. But I have some fix comments

**Comment by denys-medynskyi** [comment]
@sveredyuk ready for the next round of review 🥊

File: app/services/visit_request/check_in.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+          :not_found
+        elsif visit_request.visited?
+         :already_visited
+        elsif  !visit_request.approved?
+          :invalid_status
+        else
+         :success
+      end
+
+      visit_request.update(visited: true, checked_in_at: Time.zone.now) if status == :success
```
Comment (sveredyuk) [inline app/services/visit_request/check_in.rb]: `Time.current` always use UTC time zone.

File: app/views/admin/visit_request/check_in/_find_visitor_search.slim [hunk 6]
```diff
@@ -0,0 +1,3 @@
+p
+  b= t('visit_request.check_in.search')
```
Comment (sveredyuk) [inline app/views/admin/visit_request/check_in/_find_visitor_search.slim]: `b = t(...` space after `b`

**Review by sveredyuk — COMMENTED** [review]
Well done. Some small fixes & done

**Review by sveredyuk — DISMISSED** [review]

File: app/services/visit_request/generate_qr_code.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    end
+
+    private
+
+    attr_reader :visit_request
+
+    def qr_code
+      check_in_visit_request_url = admin_visit_request_check_in_url(token: visit_request.token, host: host)
+      # docs - https://github.com/jabbrwcky/prawn-qrcode
+      qr_code = RQRCode::QRCode.new(check_in_visit_request_url, level: :h, size: 30)
```
Comment (floor114) [inline app/services/visit_request/generate_qr_code.rb]: Useless local variable `qr_code`

**Review by floor114 — DISMISSED** [review]

**Comment by denys-medynskyi** [comment]
@sveredyuk @floor114 after I pushed force your reviews were dismissed.
Please give your approval again 🙏

**Review by floor114 — APPROVED** [review]
Don't forget to squash commits.

**Review by sveredyuk — APPROVED** [review]

**Comment by sveredyuk** [comment]
Squash & Merge

**Pull request merged**